wgengine/monitor/monitor_windows.go

```diff
@@ -51,229 +33,136 @@ type messageOrError struct {
-		if err == errClosed {
-			return
-		}
-		select {
-		case m.messagec <- messageOrError{msg, err}:
-		case <-m.ctx.Done():
-			return
+	if m.addressChangeCallback != nil {
+		if err := m.addressChangeCallback.Unregister(); err != nil {
+			m.logf("addressChangeCallback.Unregister error: %v", err)
+			// TODO: The same question about errors...
+			return err
 		}
+		m.addressChangeCallback = nil
 	}
+
+	return nil
 }
 
-func (m *winMon) getIPOrRouteChangeMessage() (message, error) {
-	if m.ctx.Err() != nil {
+func (m *winMon) Receive() (message, error) {
+	m.mu.Lock()
+
+	if m.addressChangeCallback == nil || m.routeChangeCallback == nil {
+		m.mu.Unlock()
+		m.logf("Receive() call on closed monitor")
 		return nil, errClosed
 	}
 
-	// TODO(bradfitz): locking ourselves to an OS thread here
-	// likely isn't necessary, but also can't really hurt.
-	// We'll be blocked in windows.WaitForMultipleObjects below
-	// anyway, so might as well stay on this thread during the
-	// notify calls and cancel funcs.
-	// Given the past memory corruption from misuse of these APIs,
-	// and my continued lack of understanding of Windows APIs,
-	// I'll be paranoid. But perhaps we can remove this once
-	// we understand more.
-	runtime.LockOSThread()
-	defer runtime.UnlockOSThread()
-
-	addrHandle, oaddr, cancel, err := notifyAddrChange()
+	// TODO: In the original version we haven't took the initial state, which looks as bug, but check anyway
+	prevState, err := interfaces.GetState()
 	if err != nil {
-		m.logf("notifyAddrChange: %v", err)
+		m.mu.Unlock()
+		m.logf("interfaces.GetState error: %v", err)
 		return nil, err
 	}
-	defer cancel()
 
-	routeHandle, oroute, cancel, err := notifyRouteChange()
-	if err != nil {
-		m.logf("notifyRouteChange: %v", err)
-		return nil, err
-	}
-	defer cancel()
+	m.messagec = make(chan messageOrError)
+	defer func() {
+		m.mu.Lock()
+		m.messagec = nil
+		m.mu.Unlock()
+	}()
 
-	t0 := time.Now()
-	eventNum, err := windows.WaitForMultipleObjects([]windows.Handle{
-		m.closeHandle, // eventNum 0
-		addrHandle,    // eventNum 1
-		routeHandle,   // eventNum 2
-	}, false, windows.INFINITE)
-	if m.ctx.Err() != nil || (err == nil && eventNum == 0) {
-		return nil, errClosed
-	}
-	if err != nil {
-		m.logf("waitForMultipleObjects: %v", err)
-		return nil, err
-	}
+	m.mu.Unlock()
 
-	d := time.Since(t0)
-	var eventStr string
+	// TODO: Recheck if we actually need this poller
+	pollTicker := time.NewTicker(pollInterval)
+	defer pollTicker.Stop()
 
-	// notifyAddrChange and notifyRouteChange both seem to return the same
-	// handle value. Determine which fired by looking at the "Internal" (sic)
-	// field of the Ovelapped instead.
-	// TODO(bradfitz): maybe clean this up; see TODO in callNotifyProc.
-	if (eventNum == 1 || eventNum == 2) && addrHandle == routeHandle {
-		if oaddr.Internal == _STATUS_WAIT_0 && oroute.Internal == _STATUS_PENDING {
-			eventStr = "addr-o" // "-o" overlapped suffix to distinguish from "addr" below
```

Comment:
this log info was lost. Maybe instead at the two points where do call `somethingChanged`, log some info about what the callback said was changed?

Reply:
Ah, ok. Will add `route` or `addr`...